research/fed-sm/job_configs/fedsm_prostate/app/custom/learners/supervised_learner.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+
+import numpy as np
+import torch
+from torch.utils.tensorboard import SummaryWriter
+
+from nvflare.apis.dxo import DXO, DataKind, MetaKey, from_shareable
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.learner_spec import Learner
+from nvflare.app_common.app_constant import AppConstants, ValidateType
+
+
+class SupervisedLearner(Learner):
+    def __init__(
+        self,
+        aggregation_epochs: int = 1,
+        train_task_name: str = AppConstants.TASK_TRAIN,
+    ):
+        """Simple Supervised Trainer.
+            This provides the basic functionality of a local learner: perform before-train validation on
+            global model at the beginning of each round, perform local training, and send the updated weights.
+            No model will be saved locally, tensorboard record for local loss and global model validation score.
+            Enabled FedAvg
+
+        Args:
+            train_config_filename: directory of config_3 file.
+            aggregation_epochs: the number of training epochs for a round. Defaults to 1.
+            train_task_name: name of the task to train the model.
+
+        Returns:
+            a Shareable with the updated local model after running `execute()`
+        """
+        super().__init__()
+        # trainer init happens at the very beginning, only the basic info regarding the trainer is set here
+        # the actual run has not started at this point
+        self.aggregation_epochs = aggregation_epochs
+        self.train_task_name = train_task_name
+        self.best_metric = 0.0
+
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        # when a run starts, this is where the actual settings get initialized for trainer
+
+        # set the paths according to fl_ctx
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+
+        # get and print the args
+        fl_args = fl_ctx.get_prop(FLContextKey.ARGS)
+        self.client_id = fl_ctx.get_identity_name()
```

Comment:
`self.client_id` and `self.writer` should be initialized as None in `__init__()`

Reply:
added